### PR TITLE
Evaluate live setup.py with a new --insecure option #62

### PR DESCRIPTION
**Pull request by TG1999**
Reference: https://github.com/nexB/python-inspector/issues/62
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks... See some feedback FYI.

File: src/python_inspector/resolution.py [hunk 1]
```diff
@@ -116,24 +138,51 @@ def get_environment_marker_from_environment(environment):
...
 
-def is_requirements_file_in_setup_files(setup_files: List[str]) -> bool:
+def contain_string(string: str, files: List) -> bool:
     """
-    Return True if the string ``requirements.txt`` is found in any of the ``setup_files`` location
-    strings to either a setup.py or setup.cfg file.
-    This is an indication that a requirements.txt is likely loaded in the setup.py
-    code and we use this as a hint to treat requirements.txt requirements
-    as being for the setup.py file.
+    Return True if the string is contains in any of the files.
```
Comment (pombredanne) [inline src/python_inspector/resolution.py:143]:
```suggestion
    Return True if the ``string`` is contained in any of the ``files`` list of file paths.
```

File: src/python_inspector/resolution.py [hunk 2]
```diff
@@ -116,24 +138,51 @@ def get_environment_marker_from_environment(environment):
...
             # TODO also consider other file names
-            if "requirements.txt" in f.read():
+            if string in f.read():
                 return True
     return False
 
 
+def parse_reqs_from_setup_py_insecurely(setup_py):
+    """
+    Yield requirements from the setup.py file at ``setup_py``.
```
Comment (pombredanne) [inline src/python_inspector/resolution.py:157]:
```suggestion
    Yield  Requirement(s) from a ``setup_py`` setup.py file location .
```

File: src/python_inspector/resolution.py [hunk 3]
```diff
@@ -116,24 +138,51 @@ def get_environment_marker_from_environment(environment):
...
+    """
+    if not os.path.exists(setup_py):
+        return []
+    for req in iter_requirements(level="", extras=[], setup_file=setup_py):
+        yield Requirement(req)
+
+
+def parse_deps_from_setup_py_insecurely(setup_py):
+    """
+    Yield requirements from the setup.py file at ``setup_py``.
```
Comment (pombredanne) [inline src/python_inspector/resolution.py:167]:
```suggestion
    Yield DependentPackage(s) from the ``setup_py`` setup.py file location .
```

File: src/python_inspector/resolution.py [hunk 4]
```diff
@@ -116,24 +138,51 @@ def get_environment_marker_from_environment(environment):
...
+        parsed_req = Requirement(req)
+        yield DependentPackage(
+            purl=str(
+                PackageURL(
+                    type="pypi",
+                    name=parsed_req.name,
+                )
+            ),
+            extracted_requirement=req,
+            scope="install",
```
Comment (pombredanne) [inline src/python_inspector/resolution.py:181]: Are we sure we always look exclusively for install_requires for now?

File: src/python_inspector/resolution.py [hunk 5]
```diff
@@ -705,6 +666,70 @@ def get_package_list(results):
     return list(sorted(packages))
 
 
+def get_setup_dependencies(location, analyze_setup_py_insecurely=False, use_requirements=True):
+    """
+    Yield dependencies from the given setup.py and setup.cfg location.
```
Comment (pombredanne) [inline src/python_inspector/resolution.py:671]:
```suggestion
    Yield Requirement(s) from Pypi in the ``location`` directory that contains
    a setup.py and/or a setup.cfg and optionally a requirements.txt file if
    ``use_requirements`` is True and this file is used in the setup.py or setup.cfg.
    Perform an insecure live evaluation of the Python code if needed and if
    ``analyze_setup_py_insecurely`` is True.
```

File: src/python_inspector/setup_py_live_eval.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+# This file is part of Requirements-Builder
+# Copyright (C) 2015, 2016, 2017, 2018 CERN.
+#
+# Requirements-Builder is free software; you can redistribute it and/or
+# modify it under the terms of the Revised BSD License; see LICENSE
+# file for more details.
+#
+"""Generate requirements from `setup.py` and `requirements-devel.txt`."""
+
+from __future__ import absolute_import
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:12]:
```suggestion
```

File: src/python_inspector/setup_py_live_eval.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+from __future__ import print_function
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:13]:
```suggestion
```

File: src/python_inspector/setup_py_live_eval.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+
+def build_pkg_name(pkg):
+    """Build package name, including extras if present."""
+    if pkg.extras:
+        return "{0}[{1}]".format(pkg.project_name, ",".join(sorted(pkg.extras)))
+    return pkg.project_name
+
+
+def parse_pip_file(path):
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:54]: You should remove this function that we do not use and is not as good as pip-requirements-parser.

File: src/python_inspector/setup_py_live_eval.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    import configparser
+except ImportError:  # pragma: no cover
+    import ConfigParser as configparser
+
+import mock
+import pkg_resources
+import setuptools
+
+
+def parse_set(string):
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:29]: Remove this unused function

File: src/python_inspector/setup_py_live_eval.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        # in Wheel for details.
+        condition = ""
+        if ":" in e:
+            e, condition = e.split(":", 1)
+        if not e or e in extras:
+            if condition:
+                reqs = ["{0}; {1}".format(r, condition) for r in reqs]
+            install_requires.extend(reqs)
+
+    for pkg in pkg_resources.parse_requirements(install_requires):
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:133]: This likely is worth replacing by our own requirements parsing which is better (or at least use "packaging" and not pkg_resources).

File: src/python_inspector/setup_py_live_eval.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        with open(setup_file) as sf:
+            exec(sf.read(), g)
+        sys.path.pop()
+        assert g["setup"]  # silence warning about unused imports
+    # called arguments are in `mock_setup.call_args`
+    os.chdir(cd)
+    mock_args, mock_kwargs = mock_setup.call_args
+    install_requires = mock_kwargs.get("install_requires", install_requires)
+
+    requires_extras = mock_kwargs.get("extras_require", requires_extras)
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:120]: There is also the legacy "test_requires" and "setup_requires" to support

File: src/python_inspector/setup_py_live_eval.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    # change directory to setup.py path
+    with mock.patch.object(setuptools, "setup") as mock_setup:
+        sys.path.append(os.path.dirname(setup_file))
+        g = {"__file__": setup_file, "__name__": "__main__"}
+        with open(setup_file) as sf:
+            exec(sf.read(), g)
+        sys.path.pop()
+        assert g["setup"]  # silence warning about unused imports
+    # called arguments are in `mock_setup.call_args`
+    os.chdir(cd)
```
Comment (pombredanne) [inline src/python_inspector/setup_py_live_eval.py:116]: This should be in a context manager as we cannot change the cwd forever IMHO. You can use this `commoncode.command.pushd`https://github.com/nexB/commoncode/blob/9131627677d3ef171ddc472991a5c4d4a3431ee3/src/commoncode/command.py#L246

**Review by pombredanne — APPROVED** [review]
Thanks! See a few nitpickings for your consideration and then go ahead, merge and release

File: src/python_inspector/resolution.py (hunk 4 above)
Comment (TG1999) [reply src/python_inspector/resolution.py:181]: Yes!

**Pull request merged**